File: 208-implement-trie-prefix-tree/208-implement-trie-prefix-tree.cs
public class Trie {

    public class node{
        public node[] c;
        public bool end;
        public node(){
            c=new node[26];
            end=false;
        }
    }
    public node root;
    
    public Trie() {
        root=new node();
    }
    
    public void Insert(string word) {
        var temp = root;
        foreach(var item in word){
            var ind = item-'a';
            if(temp.c[ind]==null){
                temp.c[ind]=new node();
            }
            temp = temp.c[ind];
        }
        temp.end=true;
    }
    
    public bool Search(string word) {
        var a = find(word);
        return a==null?false:a.end;
    }
    
    public bool StartsWith(string prefix) {
        var a = find(prefix);
        return a!=null;
    }
    
    public node find(string word){
        var temp = root;
        foreach(var item in word){
            var ind = item-'a';
            if(temp.c[ind]==null)return null;
            temp=temp.c[ind];
        }
        return temp;
    }
}

File: 208-implement-trie-prefix-tree/208-implement-trie-prefix-tree.cpp
class TrieNode {
public:
    bool end;
    TrieNode *c[26];
    
    TrieNode() {
        end = false;
        
        for (int i = 0; i < 26; i++)
            c[i] = NULL;
    }
};

class Trie {
public:
    TrieNode* root;
    
    Trie() {
        root = new TrieNode();
    }
    
    void insert(string word) {
        int word_len = word.length();
        TrieNode *cur = root;
        
        for (int i = 0; i < word_len; i++)
        {
            int k = word[i] - 'a';
            if (cur->c[k] == NULL)
            {
                cur->c[k] = new TrieNode();
            }
            cur = cur->c[k];
        }
        cur->end = true;
    }
    
    bool search(string word) {
        int word_len = word.length();
        TrieNode *cur = root;
        for (int i = 0; i < word_len; i++)
        {
            int k = word[i] - 'a';
            cur = cur->c[k];
            if (cur == NULL) return false;
        }        
        return cur->end;
    }
    
    bool startsWith(string prefix) {
        int word_len = prefix.length();
        TrieNode *cur = root;
        
        for (int i = 0; i < word_len; i++)
        {
            int k = prefix[i] - 'a';
            cur = cur->c[k];
            
            if (cur == NULL)
                return false;
        }
        
        return true;
    }
};

/**
 * Your Trie object will be instantiated and called as such:
 * Trie* obj = new Trie();
 * obj->insert(word);
 * bool param_2 = obj->search(word);
 * bool param_3 = obj->startsWith(prefix);
 */